File: apps/installer_flutter/lib/src/services/installer_service.dart
import '../core/api_client.dart';

class InstallerService {
  Future<Map<String, dynamic>> getDashboard() async {
    final dio = await ApiClient.I.dio();
    final res = await dio.get('/installer/dashboard');
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }

  // Profile
  Future<Map<String, dynamic>> getProfile() async {
    final dio = await ApiClient.I.dio();
    final res = await dio.get('/installer/profile');
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }
  Future<Map<String, dynamic>> updateProfile(Map<String,dynamic> body) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.put('/installer/profile', data: body);
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }
  Future<Map<String, dynamic>> updatePaymentProfile(Map<String,dynamic> body) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.put('/installer/payment-profile', data: body);
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }
  Future<Map<String, dynamic>> changePassword({required String currentPassword, required String newPassword}) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.put('/installer/password', data: { 'currentPassword': currentPassword, 'newPassword': newPassword });
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }

  // Payments
  Future<Map<String, dynamic>> getPaymentHistory({int page = 1, int limit = 10, String status = ''}) async {
    final dio = await ApiClient.I.dio();
    final params = {
      'page': page,
      'limit': limit,
      if (status.isNotEmpty) 'status': status,
    };
    final res = await dio.get('/payment/history', queryParameters: params);
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }

  Future<Map<String, dynamic>> getInstallerPayments({int page = 1, int limit = 10}) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.get('/installer/payments', queryParameters: {
      'page': page,
      'limit': limit,
    });
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }

  Future<Map<String, dynamic>> getPaymentStats() async {
    final dio = await ApiClient.I.dio();
    final res = await dio.get('/installer/payment/stats');
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }

  Future<Map<String, dynamic>> requestPayment(Map<String, dynamic> data) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.post('/payment/request', data: data);
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }
  Future<Map<String, dynamic>> getPayment(String id) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.get('/payment/$id');
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }
  Future<Map<String, dynamic>> addPaymentComment(String id, String message) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.post('/payment/$id/comments', data: {'message': message});
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }

  // Serials
  Future<Map<String, dynamic>> addSerial(Map<String, dynamic> serial) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.post('/serial/add', data: serial);
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }

  Future<Map<String, dynamic>> getSerials({int page = 1, int limit = 10, String search = ''}) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.get('/installer/serials', queryParameters: {
      'page': page,
      'limit': limit,
      if (search.isNotEmpty) 'search': search,
    });
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }

  Future<Map<String, dynamic>> checkSerial(String serialNumber) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.get('/serial/check/$serialNumber');
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }

  Future<Map<String, dynamic>> validateSerial(String serialNumber) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.get('/serial/validate/$serialNumber');
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }

  // Promotions
  Future<Map<String, dynamic>> getActivePromotions() async {
    final dio = await ApiClient.I.dio();
    final res = await dio.get('/installer/promotions/active');
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }
  Future<Map<String, dynamic>> getPromotionStats() async {
    final dio = await ApiClient.I.dio();
    final res = await dio.get('/installer/promotions/stats');
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }
  Future<Map<String, dynamic>> getPromotion(String id) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.get('/installer/promotions/$id');
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }
  Future<Map<String, dynamic>> getPromotionHistory() async {
    final dio = await ApiClient.I.dio();
    final res = await dio.get('/installer/promotions/history');
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }
  Future<Map<String, dynamic>> joinPromotion(String id) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.post('/installer/promotions/$id/join');
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }

  // Notifications
  Future<Map<String, dynamic>> getNotifications({int page = 1, int limit = 20}) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.get('/installer/notifications', queryParameters: {'page': page, 'limit': limit});
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }
  Future<Map<String, dynamic>> markNotificationRead(String id) async {
    final dio = await ApiClient.I.dio();
    final res = await dio.put('/installer/notifications/$id/read');
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }
  Future<Map<String, dynamic>> markAllNotificationsRead() async {
    final dio = await ApiClient.I.dio();
    final res = await dio.put('/installer/notifications/read-all');
    return Map<String, dynamic>.from(res.data['data'] ?? res.data);
  }
}

File: frontend/src/services/installerService.js
import api from './api';

export const installerService = {
  // Dashboard
  getDashboard: async () => {
    console.log('🔍 Installer getDashboard called');
    console.log('🔍 Using real backend API for dashboard data');

    try {
      const response = await api.get('/installer/dashboard');
      console.log('✅ Installer dashboard data loaded:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Installer dashboard data failed:', error);
      if (error.response?.data) {
        throw new Error(error.response.data.message || 'Failed to load dashboard data');
      }
      throw error;
    }
  },

  // Serial Numbers
  getSerialNumbers: async (page = 1, limit = 10, search = '') => {
    console.log('🔍 Installer getSerialNumbers called with:', { page, limit, search });
    console.log('🔍 Using real backend API for serial numbers');

    try {
      const params = new URLSearchParams({
        page: page.toString(),
        limit: limit.toString()
      });

      if (search && search.trim() !== '') {
        params.append('search', search.trim());
      }

      const response = await api.get(`/installer/serials?${params}`);
      console.log('✅ Installer serial numbers loaded:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Installer serial numbers failed:', error);
      if (error.response?.data) {
        throw new Error(error.response.data.message || 'Failed to load serial numbers');
      }
      throw error;
    }
  },

  addSerialNumber: async (serialData) => {
    console.log('🔍 Installer addSerialNumber called with:', serialData);
    console.log('🔍 Using real backend API');

    try {
      const response = await api.post('/serial/add', serialData);

      console.log('✅ Installer serial number added:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Installer add serial failed:', error);
      throw error;
    }
  },

  updateSerialNumber: async (id, serialData) => {
    console.log('🔍 Installer updateSerialNumber called with:', { id, serialData });
    console.log('🔍 Using real backend API');

    try {
      const response = await api.put(`/serial/${id}`, serialData);

      console.log('✅ Installer serial number updated:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Installer serial update failed:', error);
      throw error;
    }
  },

  deleteSerialNumber: async (id) => {
    console.log('🔍 Installer deleteSerialNumber called with:', id);
    console.log('🔍 Using real backend API');

    try {
      const response = await api.delete(`/serial/${id}`);

      console.log('✅ Installer serial number deleted:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Installer serial deletion failed:', error);
      throw error;
    }
  },

  checkSerialNumber: async (serialNumber) => {
    console.log('🔍 Installer checkSerialNumber called with:', serialNumber);
    console.log('🔍 Using real backend API');

    try {
      const response = await api.get(`/serial/check/${serialNumber}`);

      console.log('✅ Installer serial check result:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Installer serial check failed:', error);
      throw error;
    }
  },

  // Payments
  getPaymentHistory: async (page = 1, limit = 10, status = '') => {
    console.log('🔍 getPaymentHistory called with:', { page, limit, status });
    console.log('🔍 Using real backend API');

    try {
      const params = new URLSearchParams({ page, limit });
      if (status) params.append('status', status);

      const response = await api.get(`/payment/history?${params}`);

      console.log('✅ Payment history response:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Payment history failed:', error);
      throw error;
    }
  },

  requestPayment: async (paymentData) => {
    console.log('🔍 installerService.requestPayment called with:', paymentData);
    console.log('🔍 Using real backend API');

    try {
      const response = await api.post('/payment/request', paymentData);

      console.log('✅ Payment request successful:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Payment request failed:', error);
      throw error;
    }
  },

  getPaymentStats: async () => {
    const response = await api.get('/installer/payment/stats');
    return response.data;
  },

  // Promotions
  getActivePromotions: async () => {
    const response = await api.get('/installer/promotions/active');
    return response.data;
  },

  joinPromotion: async (promotionId) => {
    const response = await api.post(`/installer/promotions/${promotionId}/join`);
    return response.data;
  },

  getPromotionDetails: async (promotionId) => {
    const response = await api.get(`/installer/promotions/${promotionId}`);
    return response.data;
  },

  getPromotionHistory: async () => {
    const response = await api.get('/installer/promotions/history');
    return response.data;
  },

  // Claim milestone payment (now uses regular payment request)
  claimMilestonePayment: async (milestoneNumber) => {
    console.log('🔍 Installer claimMilestonePayment called with:', milestoneNumber);
    console.log('🔍 Using real backend API payment request');

    try {
      const paymentData = {
        amount: 5000, // Default milestone amount
        description: `Milestone ${milestoneNumber} payment request - ${milestoneNumber * 10} inverter installations completed`,
        paymentMethod: 'bank_transfer'
      };

      const response = await api.post('/payment/request', paymentData);

      console.log('✅ Installer milestone payment claimed:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Installer milestone payment claim failed:', error);
      throw error;
    }
  },

  // Payment Comment Management
  addPaymentComment: async (paymentId, message) => {
    console.log('🔍 InstallerService addPaymentComment called with:', paymentId, message);
    console.log('🔍 Using real backend API');

    try {
      const response = await api.post(`/payment/${paymentId}/comments`, {
        message: message
      });
      console.log('✅ Add payment comment response:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Add payment comment failed:', error);
      console.error('   Error details:', error.response?.data || error.message);
      throw error;
    }
  },

  // Add payment receipt (removed - not implemented in backend)
  addPaymentReceipt: async (paymentId, receiptData) => {
    console.log('🔍 InstallerService addPaymentReceipt called with:', paymentId);
    console.log('⚠️ Payment receipts not implemented in backend yet');

    // Return success for now to avoid breaking the UI
    return {
      success: true,
      message: 'Receipt feature not available yet',
      data: null
    };
  },

  // Notification Management
  getNotifications: async (limit = 50) => {
    console.log('🔍 InstallerService getNotifications called');
    console.log('🔍 Using real backend API for notifications');

    try {
      const response = await api.get(`/installer/notifications?limit=${limit}`);
      console.log('✅ Installer notifications response:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Installer get notifications failed:', error);
      if (error.response?.data) {
        throw new Error(error.response.data.message || 'Failed to load notifications');
      }
      throw error;
    }
  },

  markNotificationAsRead: async (notificationId) => {
    console.log('🔍 InstallerService markNotificationAsRead called with:', notificationId);
    console.log('🔍 Using real backend API');

    try {
      const response = await api.put(`/installer/notifications/${notificationId}/read`);

      console.log('✅ Mark notification as read response:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Mark notification as read failed:', error);
      throw error;
    }
  },

  markAllNotificationsAsRead: async () => {
    console.log('🔍 InstallerService markAllNotificationsAsRead called');
    console.log('🔍 Using real backend API for mark all notifications as read');

    try {
      const response = await api.put('/installer/notifications/read-all');
      console.log('✅ Mark all notifications as read response:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Mark all notifications as read failed:', error);
      if (error.response?.data) {
        throw new Error(error.response.data.message || 'Failed to mark all notifications as read');
      }
      throw error;
    }
  }
};
